### PR TITLE
Update sdk to support rhone upgrade

### PR DESCRIPTION
**Pull request by Lbqds**
1. Remove unnecessary fields from the artifact
2. Add the `@using(methodSelector = false)` annotation to existing std interfaces
3. Add a `skipSaveArtifacts` flag in the CLI:
    1. If `skipSaveArtifacts` is enabled or the contracts have been deployed on the mainnet, we won't regenerate bytecode for those unchanged contract
    2. If the existing contract code is changed, we will regenerate bytecode for the contract
    3. If the existing contract code remains unchanged but new contracts are added, we will only regenerate bytecode for the new contracts

### CODE REVIEW

File: packages/web3/src/contract/contract.ts [hunk 1]
```diff
@@ -821,19 +862,22 @@ export class Project {
...
     const fullNodeVersion = defaultFullNodeVersion ?? (await provider.infos.getInfosVersion()).version
     const sourceFiles = await Project.loadSourceFiles(projectRootDir, contractsRootDir)
     const { errorOnWarnings, ...nodeCompilerOptions } = { ...DEFAULT_COMPILER_OPTIONS, ...compilerOptionsPartial }
     const projectArtifact = await ProjectArtifact.from(projectRootDir)
+    const changedSources = projectArtifact?.getChangedSources(sourceFiles) ?? sourceFiles
     if (
       projectArtifact === undefined ||
-      projectArtifact.needToReCompile(nodeCompilerOptions, sourceFiles, fullNodeVersion)
+      projectArtifact.needToReCompile(nodeCompilerOptions, fullNodeVersion) ||
+      changedSources.length > 0
```
Comment (h0ngcha0) [inline packages/web3/src/contract/contract.ts:877]: Q: Shall we check `skipSaveArtifacts` here? I suppose when `skipSaveArtifacts` is false and project is deployed on mainnet, we still want to potentially regenerate the code.

File: packages/web3/src/contract/contract.ts [hunk 2]
```diff
@@ -531,9 +543,32 @@ export class Project {
...
+    for (const [_, script] of this.scripts) {
+      await saveToFile(script)
+    }
+    await this.saveStructsToFile()
+    await this.saveProjectArtifact(projectRootDir, skipSaveArtifacts, changedSources)
+  }
+
+  private async saveProjectArtifact(projectRootDir: string, skipSaveArtifacts: boolean, changedSources: SourceInfo[]) {
+    if (skipSaveArtifacts) {
+      // we should not update the `codeHashDebug` if the `skipSaveArtifacts` is enabled
```
Comment (h0ngcha0) [inline packages/web3/src/contract/contract.ts:560]: Q: Shall we update `codeHashDebug` when the source code is updated even if `skipSaveArtifacts` is enabled?

Comment (Lbqds) [reply packages/web3/src/contract/contract.ts:560]: Yes, this is how it works now.

If the `skipSaveArtifacts` is true, then we will check if the contract code changed by the `changedSource`, if the contract code is changed, then we will use the new `codeHashDebug`.

File: packages/web3/src/contract/contract.ts (hunk 1 above)
Comment (Lbqds) [reply packages/web3/src/contract/contract.ts:877]: If the project is deployed on mainnet, then the `skipSaveArtifacts` should be true, we have already checked this in the CLI.

If the `skipSaveArtifacts` is true and the contract code is unchanged, then we won't regenerate bytecode, but If the contract code is changed, we will regenerate bytecode.

Comment (h0ngcha0) [reply packages/web3/src/contract/contract.ts:877]: I see, thanks!

**Review by h0ngcha0 — APPROVED** [review]
LGTM

**Review by polarker — APPROVED** [review]
LGTM

**Pull request merged**